File: Java/linked_list/problem_0206_reverse_linked_list.java
/*
 * https://leetcode.com/problems/reverse-linked-list/
 * Explanation:
 * 
 * A simple reverse LL problem
 * 
 * Time: O(n)
 * Space: O(1) for iterative, O(n) for recursive
 * 
 * Write the test case in LinkedListTest.java
 */

package linked_list;

import java.util.ArrayList;
import java.util.Arrays;
import java.util.List;

import org.junit.Test;
import static org.junit.Assert.assertTrue;

public class problem_0206_reverse_linked_list {
    public static ListNode reverseList(ListNode head) {
        ListNode prev = null;
        ListNode curr = head;
        while (curr != null) {
            ListNode next = curr.next;
            curr.next = prev;
            prev = curr;
            curr = next;
        }
        return prev;
    }

    public static ListNode reverseListRecursive(ListNode head) {
        if (head == null || head.next == null) {
            return head;
        }
        ListNode newHead = reverseListRecursive(head.next);
        head.next.next = head;
        head.next = null;
        return newHead;
    }

    @Test
    public void testReverseLinkedList() {
        ListNode node = new ListNode();
        ListNode testCase1 = node.addMany(new ListNode(1), new ArrayList<>(Arrays.asList(2, 3, 4, 5)));
        ListNode output = problem_0206_reverse_linked_list.reverseListRecursive(testCase1);
        List<Integer> expectedNodes = new ArrayList<>(Arrays.asList(5, 4, 3, 2, 1));
        assertTrue(node.getNodesInArray(output).equals(expectedNodes));

        ListNode testCase2 = node.addMany(new ListNode(1), new ArrayList<>(Arrays.asList(2)));
        output = problem_0206_reverse_linked_list.reverseList(testCase2);
        expectedNodes = new ArrayList<>(Arrays.asList(2, 1));
        assertTrue(node.getNodesInArray(output).equals(expectedNodes));

        ListNode testCase3 = null;
        output = problem_0206_reverse_linked_list.reverseList(testCase3);
        expectedNodes = new ArrayList<Integer>(Arrays.asList());
        assertTrue(node.getNodesInArray(output).equals(expectedNodes));
    }
}

File: csharp/solution_and_tests/linked_list/problem_0206_reverse_linked_list.cs
/*
 * https://leetcode.com/problems/reverse-linked-list/
 * Explanation:
 * 
 * While traversing the list, we can change the current node's next pointer to point to its previous element. 
 * Since a node does not have reference to its previous node, we must store its previous element beforehand. 
 * We also need another pointer to store the next node before changing the reference. 
 * Do not forget to return the new head reference at the end!
 * 
 * Time: O(n)
 * Space: O(1) for iterative, O(n) for recursive
 */

using List_Node;
using System.Collections.Generic;
using NUnit.Framework;

namespace Reverse_Linked_List
{
    class Solution
    {
        public static ListNode ReverseList(ListNode head)
        {
            ListNode prev = null;
            ListNode curr = head;
            while (curr != null)
            {
                ListNode next = curr.next;
                curr.next = prev;
                prev = curr;
                curr = next;
            }
            return prev;
        }

        public static ListNode ReverseListRecursive(ListNode head)
        {
            if (head == null || head.next == null)
            {
                return head;
            }
            ListNode newHead = ReverseListRecursive(head.next);
            head.next.next = head;
            head.next = null;
            return newHead;
        }
    }

    class Tests
    {
        Solution sol;
        ListNode node;
        [SetUp]
        public void SetUp()
        {
            sol = new Solution();
            node = new ListNode();
        }

        [Test]
        public void TestReverseList()
        {
            ListNode testCase1 = node.AddMany(new ListNode(1), new List<int> { 2, 3, 4, 5 });
            List<int> testCase1ExpectedResult = new List<int> { 5, 4, 3, 2, 1 };
            ListNode testCase2 = node.AddMany(new ListNode(1), new List<int> { 2 });
            List<int> testCase2ExpectedResult = new List<int> { 2, 1 };
            ListNode testCase3 = null;
            List<int> testCase3ExpectedResult = new List<int>();

            CollectionAssert.AreEqual(testCase1ExpectedResult, node.GetNodesInArray(Solution.ReverseList(testCase1)));
            CollectionAssert.AreEqual(testCase2ExpectedResult, node.GetNodesInArray(Solution.ReverseList(testCase2)));
            CollectionAssert.AreEqual(testCase3ExpectedResult, node.GetNodesInArray(Solution.ReverseList(testCase3)));
        }

        [Test]
        public void TestReverseListRecursive()
        {
            ListNode testCase1 = node.AddMany(new ListNode(1), new List<int> { 2, 3, 4, 5 });
            List<int> testCase1ExpectedResult = new List<int> { 5, 4, 3, 2, 1 };
            ListNode testCase2 = node.AddMany(new ListNode(1), new List<int> { 2 });
            List<int> testCase2ExpectedResult = new List<int> { 2, 1 };
            ListNode testCase3 = null;
            List<int> testCase3ExpectedResult = new List<int>();

            CollectionAssert.AreEqual(testCase1ExpectedResult, node.GetNodesInArray(Solution.ReverseListRecursive(testCase1)));
            CollectionAssert.AreEqual(testCase2ExpectedResult, node.GetNodesInArray(Solution.ReverseListRecursive(testCase2)));
            CollectionAssert.AreEqual(testCase3ExpectedResult, node.GetNodesInArray(Solution.ReverseListRecursive(testCase3)));
        }
    }
}
